universal-application-tool-0.0.1/app/views/LoginForm.java

```diff
@@ -12,13 +12,18 @@
   public Content render(Http.Request request) {
     return htmlContent(
         body(
+            h1("Error: You are not logged in")
+                .withCondHidden(!request.queryString("message").orElse("").equals("login")),
             h1("Log In"),
             form(
                     makeCsrfTokenInputTag(request),
                     textField("uname", "username", "Username"),
                     passwordField("pwd", "password", "Password"),
                     submitButton("login", "Submit"))
                 .withMethod("POST")
-                .withAction("/callback?client_name=FormClient")));
+                .withAction("/callback?client_name=FormClient"),
+            h1("Or, continue as guest."),
+            button("guest", "continue")
+                .attr("onclick", "window.location = '/callback?client_name=GuestClient';")));
```

Comment:
suggestion (non-blocking): this can be done without using JS

```java
import static j2html.attributes.Attr.FORMACTION;
...
form(
    button("guest", "continue")
    .attr(
        FORMACTION, route))
```

Reply:
Tried it, but it doesn't work with URL parameters.  We can get around that by making it a `POST` form, or by providing the url parameter as a hidden input, both of which seem like slightly too much workaround to avoid a JS fragment.  Plus, if we make it a form, I think we have to interface with Play's XSRF tokens, which is easy, but also seems like a little too much workaround.

This is some super standard javascript, the entire front page of a google search for "html button redirect" is all a variation on this, so I'm sure it's just as widely supported in browsers as html forms are, and I think it's marginally more readable than the formaction + hidden input + xsrf token approach.

Guessing by non-blocking you don't feel _too_ strongly and the added complexity there will be persuasive, but post again if you disagree - willing to go the formaction route if you're thinking we should go a bit further to avoid JS.

Reply:
Yep, didn't feel strongly about it. Thanks for the info though it's helpful to know!

Reply:
👍  Good to know.

Reply:
Though as Aaron points out in another PR, factoring this out would be helpful: https://github.com/seattle-uat/universal-application-tool/pull/167/files#r579390736